extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java

```diff
@@ -0,0 +1,506 @@
+package io.quarkus.hibernate.reactive.panache.common.runtime;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import javax.enterprise.util.TypeLiteral;
+import javax.persistence.LockModeType;
+import javax.persistence.PersistenceException;
+import javax.transaction.SystemException;
+import javax.transaction.TransactionManager;
+
+import org.hibernate.internal.util.LockModeConverter;
+import org.hibernate.reactive.mutiny.Mutiny;
+import org.hibernate.reactive.mutiny.Mutiny.Session;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.panache.common.Parameters;
+import io.quarkus.panache.common.Sort;
+import io.quarkus.panache.common.exception.PanacheQueryException;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+
+public abstract class AbstractJpaOperations<PanacheQueryType> {
+
+    protected abstract PanacheQueryType createPanacheQuery(Uni<Mutiny.Session> em, String query, String orderBy,
+            Object paramsArrayOrMap);
+
+    protected abstract Uni<List<?>> list(PanacheQueryType query);
+
+    protected abstract Multi<?> stream(PanacheQueryType query);
+
+    //
+    // Instance methods
+
+    public Uni<Void> persist(Object entity) {
+        return getEntityManager().flatMap(em -> persist(em, entity));
+    }
+
+    public Uni<Void> persist(Mutiny.Session em, Object entity) {
+        if (!em.contains(entity)) {
+            return em.persist(entity).map(v -> null);
+        }
+        return Uni.createFrom().nullItem();
+    }
+
+    public Uni<Void> persist(Iterable<?> entities) {
+        return persist(StreamSupport.stream(entities.spliterator(), false));
+    }
+
+    public Uni<Void> persist(Object firstEntity, Object... entities) {
+        List<Object> array = new ArrayList<>(entities.length + 1);
+        array.add(firstEntity);
+        for (Object entity : entities) {
+            array.add(entity);
+        }
+        return persist(array.stream());
+    }
+
+    public Uni<Void> persist(Stream<?> entities) {
+        return getEntityManager().flatMap(
+                em -> Uni.combine().all().unis(entities.map(entity -> persist(em, entity)).collect(Collectors.toList()))
+                        .combinedWith(list -> null));
+    }
+
+    public Uni<Void> delete(Object entity) {
+        return getEntityManager().flatMap(em -> em.remove(entity).map(v -> null));
+    }
+
+    public Uni<Boolean> isPersistent(Object entity) {
+        return getEntityManager().map(em -> em.contains(entity));
+    }
+
+    public Uni<Void> flush() {
+        return getEntityManager().flatMap(em -> em.flush()).map(v -> null);
+    }
+
+    //
+    // Private stuff
+
+    public Uni<Mutiny.Session> getEntityManager() {
```

Comment:
why name this "getEntityManager" when it doesn't return one? :)

Also: why returning an Uni<Session>, when you could just get the `Session` right away? opening a Session is non-blocking.

Reply:
I mean, ask ArC for a `Mutiny.Session` rather than a `Uni<Mutiny.Session>`

Reply:
I don't think there's any producer for `Mutiny.Session` since ArC doesn't do async injection, so not sure what you're talking about.

OK, apparently this was added after I did my PR. Fine, let's switch to that :)

Reply:
BTW, opening a Session isn't a blocking operation.